### PR TITLE
Add `DistributedNeighborLoader` [3/6]

### PR DESCRIPTION
**Pull request by JakubPietrakIntel**
**[2/3] Distributed Loaders PRs**
This PR includes`DistributedNeighborLoader` used for processing node sampler output in distributed training setup.

1.  https://github.com/pyg-team/pytorch_geometric/pull/8079
2.  https://github.com/pyg-team/pytorch_geometric/pull/8080
3.  https://github.com/pyg-team/pytorch_geometric/pull/8085

Other PRs related to this module:
DistSampler: https://github.com/pyg-team/pytorch_geometric/pull/7974
GraphStore\FeatureStore: https://github.com/pyg-team/pytorch_geometric/pull/8083

### CODE REVIEW

File: test/distributed/test_dist_neighbor_loader.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+        rpc_worker_names={},
+        concurrency=10,
+        device=device,
+        drop_last=True,
+        async_sampling=async_sampling,
+    )
+
+    edge_index = part_data[1]._edge_index[(None, "coo")]
+
+    assert "DistNeighborLoader()" in str(loader)
```
Comment (kgajdamo) [inline test/distributed/test_dist_neighbor_loader.py:94]: assert "DistNeighborLoader" in str(loader)

**Comment by kgajdamo** [comment]
please add DistNeighborLoader to the distributed/__init__.py file

File: test/distributed/test_dist_neighbor_loader.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    w0.join()
+    w1.join()
+
+
+@onlyLinux
+@withPackage('pyg_lib')
+@pytest.mark.parametrize('num_parts', [2])
+@pytest.mark.parametrize('num_workers', [0])
+@pytest.mark.parametrize('async_sampling', [True])
+@pytest.mark.skip(reason="Breaks with no attribute 'num_hops'")
```
Comment (rusty1s) [inline test/distributed/test_dist_neighbor_loader.py:226]: @JakubPietrakIntel This test currently breaks for me. I am disabling it for now.

**Review by rusty1s — APPROVED** [review]

File: torch_geometric/loader/node_loader.py [hunk 3]
```diff
@@ -160,25 +152,19 @@ def filter_fn(
...
-                    out.col,
-                    out.edge,
-                    self.node_sampler.edge_permutation,
-                )
+                data = filter_data(  #
+                    self.data, out.node, out.row, out.col, out.edge,
+                    self.node_sampler.edge_permutation)
+
             else:  # Tuple[FeatureStore, GraphStore]
-                data = Data(
```
Comment (kgajdamo) [inline torch_geometric/loader/node_loader.py]: @rusty1s I think this code was to add the features to the data.

Comment (rusty1s) [reply torch_geometric/loader/node_loader.py]: Where was `metadata[-3]` and `metadata[-2]` populated? The sampler does not have access to node features, so I got confused by this part and dropped it.

Comment (kgajdamo) [reply torch_geometric/loader/node_loader.py]: This info is added to the SamplerOutput metadata after sampling and collecting features from all machines in the:
[DistSampler _collate()](https://github.com/pyg-team/pytorch_geometric/blob/9f7e82442efb41687e60469f1e63b0efaa793165/torch_geometric/distributed/dist_neighbor_sampler.py#L717C1-L717C1)

Comment (JakubPietrakIntel) [reply torch_geometric/loader/node_loader.py]: The output of `collate_fn()` includes `nfeats, nlabels, efeats` which are our `metadata[-3]`, `metadata[-2]` and `metadata[-1]`. Here: [L717C1-L718C22](https://github.com/pyg-team/pytorch_geometric/blob/88d7986b6d0a6de5895872270d2ff4fc95fae3b7/torch_geometric/distributed/dist_neighbor_sampler.py#L717C1-L718C22)
I will add a small PR to revert these lines.

Comment (ZhengHongming888) [reply torch_geometric/loader/node_loader.py]: @rusty1s It is different from single node case. for distributed after node sampling there will need one step to put labels, nfeats, efeats from different nodes into pyg data format for loader.
